### PR TITLE
Fixed exception with users without email address

### PR DESCRIPTION
**Pull request by Eldblom**

### CODE REVIEW

**Comment by kamsar** [comment]
Looks good; will merge with next release

**Comment by Eldblom** [comment]
Just found a corresponding error going the other way

**Comment by Eldblom** [comment]
Fixes #190

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 1]
```diff
@@ -39,9 +37,7 @@ public void Load(IConfiguration configuration)
 		{
 			using (new UnicornOperationContext())
 			{
-				var users = _userDataStore
-					.GetAll()
-					.Where(user => _userPredicate.Includes(User.FromName(user.User.UserName, false)).IsIncluded).ToArray();
+				var users = _userDataStore.GetAll().Where(user => _userPredicate.Includes(User.FromName(user.User.UserName, false)).IsIncluded).ToArray();
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 2]
```diff
@@ -50,10 +46,7 @@ public void Load(IConfiguration configuration)
 
 				if (_syncConfiguration.RemoveOrphans)
 				{
-					var existingOrphanUsers = UserManager.GetUsers()
-						.GetAll()
-						.Where(user => _userPredicate.Includes(user).IsIncluded)
-						.Where(includedUser => !users.Any(user => user.User.UserName.Equals(includedUser.Name, StringComparison.OrdinalIgnoreCase)));
+					var existingOrphanUsers = UserManager.GetUsers().GetAll().Where(user => _userPredicate.Includes(user).IsIncluded).Where(includedUser => !users.Any(user => user.User.UserName.Equals(includedUser.Name, StringComparison.OrdinalIgnoreCase)));
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 3]
```diff
@@ -81,27 +72,18 @@ protected virtual void DeserializeUser(SyncUserFile serializedUser)
...
-				null,
-				syncUser.Comment,
-				syncUser.IsApproved,
-				existingUser.IsLockedOut,
-				syncUser.CreationDate,
-				existingUser.LastLoginDate,
-				existingUser.LastActivityDate,
-				addedUser ? DateTime.UtcNow : existingUser.LastPasswordChangedDate,
-				existingUser.LastLockoutDate);
+			var updatedUser = new MembershipUser(existingUser.ProviderName, syncUser.UserName, existingUser.ProviderUserKey, syncUser.Email, null, syncUser.Comment, syncUser.IsApproved, existingUser.IsLockedOut, syncUser.CreationDate, existingUser.LastLoginDate, existingUser.LastActivityDate, addedUser ? DateTime.UtcNow : existingUser.LastPasswordChangedDate, existingUser.LastLockoutDate);
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 4]
```diff
@@ -81,27 +72,18 @@ protected virtual void DeserializeUser(SyncUserFile serializedUser)
...
-				if (!existingUser.CreationDate.Equals(updatedUser.CreationDate.ToUniversalTime())) changes.Add(new UserUpdate("CreationDate", existingUser.CreationDate.ToString(CultureInfo.InvariantCulture), updatedUser.CreationDate.ToUniversalTime().ToString(CultureInfo.InvariantCulture) + " (UTC)"));
+				if (existingUser.Email != null && !existingUser.Email.Equals(updatedUser.Email))
+					changes.Add(new UserUpdate("Email", existingUser.Email, updatedUser.Email));
+				if (existingUser.Comment != null && !existingUser.Comment.Equals(updatedUser.Comment))
+					changes.Add(new UserUpdate("Comment", existingUser.Comment, updatedUser.Comment));
+				if (!existingUser.IsApproved.Equals(updatedUser.IsApproved))
+					changes.Add(new UserUpdate("IsApproved", existingUser.IsApproved.ToString(), updatedUser.IsApproved.ToString()));
+				if (!existingUser.CreationDate.Equals(updatedUser.CreationDate.ToUniversalTime()))
+					changes.Add(new UserUpdate("CreationDate", existingUser.CreationDate.ToString(CultureInfo.InvariantCulture), updatedUser.CreationDate.ToUniversalTime().ToString(CultureInfo.InvariantCulture) + " (UTC)"));
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: For multi-line `if` statements please add curly braces (e.g.

```
if(foo)
{
 // do stuff
}
```

Or else return the formatting to how it was.

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 5]
```diff
@@ -120,7 +102,8 @@ protected virtual MembershipUser GetOrCreateUser(SyncUserFile user, out bool add
 
 			var existingUser = Membership.GetUser(user.User.UserName);
 
-			if (existingUser != null) return existingUser;
+			if (existingUser != null)
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 6]
```diff
@@ -137,22 +120,23 @@ protected virtual string CreateNewUserPassword(SyncUserFile user)
 				return Membership.GeneratePassword(32, 0);
 			}
 
-			if (password.Length < 8) throw new InvalidOperationException("I will not set the default user password to anything less than 8 characters. Change your userSyncConfiguration's defaultPassword setting to something more secure.");
+			if (password.Length < _syncConfiguration.MinPasswordLength)
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 7]
```diff
@@ -162,11 +146,12 @@ protected virtual void PasteProfileValues(MembershipUser updatedUser, SyncUser s
 
 				// check if we need to change the value
 				var existingValue = user.Profile.GetCustomProperty(customProperty.Name);
-				if (existingValue != null && existingValue.Equals((string) customProperty.Content, StringComparison.Ordinal)) continue;
+				if (existingValue != null && existingValue.Equals((string) customProperty.Content, StringComparison.Ordinal))
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please add curly braces for readability

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 8]
```diff
@@ -188,7 +173,8 @@ protected virtual void PasteProfileValues(MembershipUser updatedUser, SyncUser s
 
 				if (existingValue != null && (existingValue.GetType().IsPrimitive || existingValue is string))
 				{
-					if (existingValue.Equals(standardProperty.Content)) continue; // no changes, skip
+					if (existingValue.Equals(standardProperty.Content))
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 9]
```diff
@@ -200,7 +186,8 @@ protected virtual void PasteProfileValues(MembershipUser updatedUser, SyncUser s
 					// we'll go with a quiet always update here
 					propertiesAreUpdated = true;
 					user.Profile.SetPropertyValue(standardProperty.Name, standardProperty.Content);
-					if (existingValue == null) changes.Add(new UserUpdate(standardProperty.Name, "null", standardProperty.Content.ToString()));
+					if (existingValue == null)
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 10]
```diff
@@ -221,15 +208,17 @@ protected virtual void PasteRoles(SyncUser serializedUser, User sitecoreUser, bo
 				if (!sitecoreUser.Roles.Any(ur => ur.Name.Equals(role, StringComparison.OrdinalIgnoreCase)))
 				{
 					sitecoreUser.Roles.Add(Role.FromName(role));
-					if (!addedUser) changes.Add(new UserRoleUpdate(role, false));
+					if (!addedUser)
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Loader/UserLoader.cs [hunk 11]
```diff
@@ -221,15 +208,17 @@ protected virtual void PasteRoles(SyncUser serializedUser, User sitecoreUser, bo
...
+						changes.Add(new UserRoleUpdate(role, false));
 				}
 			}
 
 			foreach (var orphanRole in sitecoreUser.Roles.Where(role => !serializedUser.Roles.Any(sr => sr.Equals(role.Name, StringComparison.OrdinalIgnoreCase))))
 			{
 				sitecoreUser.Roles.Remove(orphanRole);
-				if (!addedUser) changes.Add(new UserRoleUpdate(orphanRole.Name, true));
+				if (!addedUser)
```
Comment (kamsar) [inline src/Unicorn.Users/Loader/UserLoader.cs]: Please preserve existing formatting

File: src/Unicorn.Users/Standard Config Files/Unicorn.Users.config [hunk 12]
```diff
@@ -17,7 +17,7 @@
 				<!-- Note: do not colocate serialized items and serialized users in the same folder -->
 				<userDataStore physicalRootPath="$(dataFolder)\Unicorn\Unicorn.Users\$(configurationName)" type="Unicorn.Users.Data.FilesystemUserDataStore, Unicorn.Users" singleInstance="true"/>
 				<userLoader type="Unicorn.Users.Loader.UserLoader, Unicorn.Users" singleInstance="true" />
-					<userLoaderLogger type="Unicorn.Users.Loader.DefaultUserLoaderLogger, Unicorn.Users" singleInstance="true" />
+				<userLoaderLogger type="Unicorn.Users.Loader.DefaultUserLoaderLogger, Unicorn.Users" singleInstance="true" />
```
Comment (kamsar) [inline src/Unicorn.Users/Standard Config Files/Unicorn.Users.config]: Please preserve existing formatting

**Comment by kamsar** [comment]
I have manually made the formatting changes requested and "merged" this with Beyond Compare. This will be in the next version.

Thanks for the contribution.

**Pull request closed without merging**